src/views/figure_editable.ts

```diff
@@ -50,11 +60,160 @@ export class FigureEditableView extends FigureView {
+        if (parent) {
+          Array.from(parent.children).forEach((el) => {
+            if (el !== this.container && el.classList.contains('figure')) {
+              el.classList.add('drag-active')
+            }
+          })
+        }
+      })
+    }
+
+    this.container.addEventListener('dragend', () => {
+      this.isDragging = false
+      figureDragState.clearActiveDrag()
+      this.container.classList.remove('dragging')
+      const parent = this.container.parentElement
+      if (parent) {
+        Array.from(parent.children).forEach((el) => {
+          el.classList.remove(
+            'drag-active',
+            'drop-target-above',
+            'drop-target-below'
+          )
+        })
+      }
+    })
+
+    this.container.addEventListener('dragover', (e) => {
+      if (!this.isDragging) {
+        e.preventDefault()
+        const rect = this.container.getBoundingClientRect()
+        const relativeY = e.clientY - rect.top
+        const isAbove = relativeY < rect.height / 2
+        this.container.classList.remove(  // Add a blue dotted border to indicate the drop zone
+          'drop-target-above',
+          'drop-target-below'
+        )
+        this.container.classList.add(
+          isAbove ? 'drop-target-above' : 'drop-target-below'
+        )
+      }
+    })
+
+    this.container.addEventListener('dragleave', (e) => {
+      if (!this.container.contains(e.relatedTarget as Node)) {
+        this.container.classList.remove(
+          'drop-target-above',
+          'drop-target-below'
+        )
+      }
+    })
+
+    this.container.addEventListener('drop', (e) => {
+      e.preventDefault()
+      e.stopPropagation()
+      const activeDrag = figureDragState.getActiveDrag()
+      let fromPos: number | null = null
+      let figureId: string | null = null
+      if (activeDrag) {
+        fromPos = activeDrag.pos
+        figureId = activeDrag.id
+      } else {
+        figureId = e.dataTransfer?.getData('text/plain') || null
+        if (figureId) {
+          fromPos = this.findFigurePosition(figureId)
+        }
+      }
+      if (fromPos === null) {
+        return
+      }
+      const toPos = this.getPos()
+      if (fromPos === toPos) {
+        return
+      } // prevent self-move
+      this.moveFigure(fromPos, toPos)
+      this.container.classList.remove('drop-target-above', 'drop-target-below')
+    })
+  }
+
+  private findFigurePosition(figureId: string): number | null {
+    let foundPos: number | null = null
+    this.view.state.doc.descendants((node, pos) => {
+      if (node.type === schema.nodes.figure && node.attrs.id === figureId) {
+        foundPos = pos
+        return false
+      }
+    })
+    return foundPos
+  }
+
+  private moveFigure(fromPos: number, targetPos: number) {
+    const { state } = this.view
+    const { tr } = state
+    const fromNode = state.doc.nodeAt(fromPos)
+    if (!fromNode) {
+      return
+    }
+    if (fromNode.type !== schema.nodes.figure) {
```

Comment:
I think 'fromPos' is found by 'findFigurePosition' so that's guaranteed to be a figure and this check is, then, not needed, right?